### ISSUE TITLE
Reference tracking using NDI Polaris Vega ST retirns global coordinates

### ISSUE THREAD

**Issue by Mil228** [PlusToolkit/PlusLib#1111]
Hi all,

I have been struggling with an issue for quite a while and cannot seem to fix it myself. I am in need of your help! While using my config file in fCal, I seem to be able to connect to a Philips Epiq Ultrasound machine through DNL services and to a NDI Polaris Vega ST Optical tracker. My desired output would be an MHA file containing the transformation matrix from tool 449 (id='Geometry') to tool 339 (id='Reference'). I set my ToolReferenceFrame to 'Reference' and changed a few options but the tracking results in the MHA file () are still in global ('OpticalTracker') coordinates instead of relative tracked to the slightly moving 'Reference' tool. What am I doing wrong? Do I need to change the xml config completely? Note that I am not using any PlusOpenIGTLink output, I just need to be able to press record and save the relative tracking data to a MHA file. I also do not need the ImageToGeometry transform as I import my tracking data as well as the DNL data into Matlab and perform the transformations there. But the realtive tracking data is essential for this. Please, is there anyone that could help me? I would very much appreciate it!!

[example_mha.txt](https://github.com/PlusToolkit/PlusLib/files/12584355/example_mha.txt)
[example_config.txt](https://github.com/PlusToolkit/PlusLib/files/12584365/example_config.txt)

**Comment by Mil228** [comment]
My log file gives me this (after changing the ToolReferenceFrame to Tracker):

|DEBUG|1101.454000| Application configuration file 'P:/PlusApp-2.8.0.20191105-Win64/bin/PlusConfig.xml' saved| in P:\workspace2\PlusLib\src\PlusCommon\vtkPlusConfig.cxx(536)
|DEBUG|1101.471000| Device set configuration is read from file: C:\Users\s137228\Documents\OneDrive - TU Eindhoven\Config files\MEDUSA_test7.xml| in P:\workspace2\PlusLib\src\PlusCommon\vtkPlusConfig.cxx(249)
|DEBUG|1101.473000| Device set configuration file contents: | in P:\workspace2\PlusLib\src\PlusCommon\vtkPlusConfig.cxx(252)
|INFO|1101.476000| Connect to devices| in P:\workspace2\PlusApp\fCal\Toolboxes\QConfigurationToolbox.cxx(158)
|DEBUG|1101.514000| StartupDelaySec: 1.000000| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDataCollector.cxx(108)
|DEBUG|1101.516000| AveragedItemsForFiltering is not defined in source element "GeometryToTracker". Using default value: 20| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDataSource.cxx(401)
|DEBUG|1101.518000| AveragedItemsForFiltering is not defined in source element "ReferenceToTracker". Using default value: 20| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDataSource.cxx(401)
|INFO|1101.519000| OpticalTracker: Local time offset: -40ms| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1098)
|WARNING|1101.520000| DnlDevice: ToolReferenceFrame is undefined. Default "(undefined)" will be used.| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(958)
|DEBUG|1101.547000| AveragedItemsForFiltering is not defined in source element "DnlFrameToDnlDevice". Using default value: 20| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDataSource.cxx(401)
|INFO|1101.547000| DnlDevice: Local time offset: -40ms| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1098)
|DEBUG|1101.548000| TrackedVideoDevice: Unable to find acquisition rate in device element when it is required, using default 50| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1050)
|DEBUG|1101.548000| TrackedVideoDevice: Local time offset was not defined in device configuration| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1103)
|DEBUG|1101.548000| OpticalTracker: vtkPlusDevice::Connect: OpticalTracker| in P:\workspace2\PlusLib\src\PlusDataCollection\vtkPlusDevice.cxx(1226)
|DEBUG|1102.574000| NDI Command:INIT| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(246)
|DEBUG|1102.721000| NDI Reply: OKAY| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(276)
|DEBUG|1102.722000| NDI Command:PHSR:01| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(246)
|DEBUG|1102.724000| NDI Reply: 00| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(276)
|DEBUG|1102.725000| NDI Command:PHRQ:*********1****| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(246)
|DEBUG|1102.728000| NDI Reply: 01| in P:\workspace2\PlusLib\src\PlusDataCollection\NDICAPITracking\vtkPlusNDITracker.cxx(276)


It shows that for example “AveragedItemsForFiltering is not defined in source element "ReferenceToTracker". “ is not given, where could I add this? I tried multiple places within the config file but no luck yet. Furthermore, it shows that it will compute the GeometryToTRacker and ReferenceToTracker but why not the desired GeometryToReference?

**Comment by adamrankin** [comment]
You can change AveragedItemsForFiltering but it's not necessary. I would reduce your log level to INFO to make it easier to read and to filter out the unimportant entries.

I don't see where it says it won't compute GeometryToReference

**Comment by Mil228** [comment]
Thank you, I will change it to info! Regarding the ‘GeometryToReference’, My .MHA file only returns “GeometryToTracker” and “ReferenceToTracker”, where can I add a tag in the config XML to also have my “GeometryToReference” returned? I am also fine with just the latter being returned and not having the global coordinates. I really need those relative positions for my tracking pipeline!

**Comment by adamrankin** [comment]
In the `<TransformNames>` tag you choose which transforms to send to a client.

**Comment by Mil228** [comment]
I have put the desired Transforms in the TransformNames tag as you can see here:

  <PlusOpenIGTLinkServer
    MaxNumberOfIgtlMessagesToSend="1"
    MaxTimeSpentWithProcessingMs="50"
    ListeningPort="18944"
    SendValidTransformsOnly="TRUE"
    OutputChannelId="TrackerStream" >
    <DefaultClientInfo>
      <MessageTypes>
        <Message Type="TRANSFORM" />
      </MessageTypes>
      <TransformNames>
                               <Transform Name="GeometryToReference" />
              <Transform Name="GeometryToTracker" />
                               <Transform Name="ReferenceToTracker" />
      </TransformNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>

But when I press Record in the Capturing tab of fCal, the recorded .MHA file (the output file generated by recording) only contains GeometryToTracker and ReferenceToTracker. What am I doing wrong?

**Comment by adamrankin** [comment]
Are there any warnings or errors in the log for that run?

**Comment by Mil228** [comment]
No, not on this part at least. I got some rendering warnings/errors but that is because I am not trying to render anything and did not include that in the config file. I only want the GeometryToReference to appear in my capturing output. I have been looking at the file format here:

http://perk-software.cs.queensu.ca/plus/doc/nightly/user/FileSequenceFile.html

But haven't found what I have been looking for... Is there no transform that I can id in the device or somewhere to get this transformation to appear in my MHA (sequence output) file?

**Comment by adamrankin** [comment]
I had not noticed you were using fCal. What channel have you selected when you choose to record?

**Comment by Mil228** [comment]
Sorry for not closing this thread earlier, but I have found the solution by using this:
http://perk-software.cs.queensu.ca/plus/doc/nightly/user/ApplicationEditSequenceFile.html

I will close the thread now!
